src/daft-file/src/python.rs

```diff
@@ -210,89 +285,195 @@ impl ObjectSourceReader {
             uri,
             position: 0,
             io_stats,
+            cached_content: None,
+            supports_range: None,
         }
     }
+    // Helper to read the entire file content
+    fn read_full_content(&self) -> io::Result<Vec<u8>> {
+        let rt = common_runtime::get_io_runtime(true);
+
+        let source = self.source.clone();
+        let uri = self.uri.clone();
+        let io_stats = self.io_stats.clone();
+
+        rt.block_within_async_context(async move {
+            let result = source
+                .get(&uri, None, io_stats)
+                .await
+                .map_err(map_get_error)?;
+
+            result
+                .bytes()
+                .await
+                .map(|b| b.to_vec())
+                .map_err(map_bytes_error)
+        })
+        .map_err(map_async_error)
+        .flatten()
+    }
 }
 
 // Implement Read for synchronous reading
+
 impl Read for ObjectSourceReader {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         if buf.is_empty() {
             return Ok(0);
         }
 
+        // If we have cached content, serve from cache
+        if let Some(ref content) = self.cached_content {
+            if self.position >= content.len() {
+                return Ok(0); // EOF
+            }
+
+            let available = content.len() - self.position;
+            let bytes_to_read = std::cmp::min(buf.len(), available);
+
+            buf[..bytes_to_read]
+                .copy_from_slice(&content[self.position..self.position + bytes_to_read]);
+            self.position += bytes_to_read;
+
+            return Ok(bytes_to_read);
+        }
+
+        // First time reading, or range support is known
         let rt = common_runtime::get_io_runtime(true);
         let start = self.position;
         let end = start + buf.len();
-        let range = Some(GetRange::Bounded(start..end));
 
+        // If we already know range requests aren't supported, read full content
+        if self.supports_range == Some(false) {
+            // Read entire file and cache it
+            let content = self.read_full_content()?;
```

Comment:
admittedly, I'm a bit on the fence about this. I think it could cause some unexpected memory usage. But it does make the api easier to use. 

I've been thinking about if this should be configurable, or opt-in/out. 


something like
```py
daft.set_execution_config(file_on_unsupported_range_request="download" | "error")
```